lib/eigensolve_quda.cpp

```diff
@@ -214,6 +214,94 @@ namespace quda
     return sum;
   }
 
+  void EigenSolver::permuteVecs(std::vector<ColorSpinorField *> &kSpace, int *mat, int size)
+  {
+    std::vector<int> pivots(size);
+    for (int i = 0; i < size; i++) {
+      for (int j = 0; j < size; j++) {
+        if (mat[j * size + i] == 1) { pivots[j] = i; }
+      }
+    }
+
+    // Identify cycles in the permutation array.
+    // We shall use the sign bit as a marker. If the
+    // sign is negative, the vector has already been
+    // swapped into the correct place. A positive
+    // value indicates the start of a new cycle.
+
+    for (int i = 0; i < size; i++) {
+      // First cycle always starts at 0, hence OR statement
+      if (pivots[i] > 0 || i == 0) {
+        int k = i;
+        // Identify vector to be placed at i
+        int j = pivots[i];
+        pivots[i] = -pivots[i];
+        while (j > i) {
+          std::swap(kSpace[k + num_locked], kSpace[j + num_locked]);
+          pivots[j] = -pivots[j];
+          k = j;
+          j = -pivots[j];
+        }
+      }
+    }
+    // Sanity check
+    for (int i = 0; i < size; i++) {
+      if (pivots[i] > 0) errorQuda("Error at pivot %d", i);
+    }
+  }
+
+  void EigenSolver::blockRotate(std::vector<ColorSpinorField *> &kSpace, double *array, int rank, int is, int ie,
+                                int js, int je, blockType b_type)
```

Comment:
I wouldn't hold this PR up on this point, but if it's easy, I liked the `range = std::pair<size_t,size_t>` @maddyscientist used in `lib/multi_blas_quda.cu`, in large part for readability.

Reply:
I'm not following this comment @weinbe2.  What does the `range` have to do with the `blockRotate` function here.  Not sure how this would apply here.

Reply:
@maddyscientist he's passing in a start and an end, a `std::pair == range` is a nice way to make that compact.

Reply:
will add this now